state/runtime/tracer/structtracer/tracer.go

```diff
@@ -0,0 +1,374 @@
+package structtracer
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+	"sync/atomic"
+
+	"github.com/0xPolygon/polygon-edge/helper/hex"
+	"github.com/0xPolygon/polygon-edge/state/runtime"
+	"github.com/0xPolygon/polygon-edge/state/runtime/evm"
+	"github.com/0xPolygon/polygon-edge/state/runtime/tracer"
+	"github.com/0xPolygon/polygon-edge/types"
+)
+
+type Config struct {
+	EnableMemory     bool // enable memory capture
+	EnableStack      bool // disable stack capture
+	EnableStorage    bool // disable storage capture
+	EnableReturnData bool // enable return data capture
+}
+
+type StructLog struct {
+	Pc            uint64                    `json:"pc"`
+	Op            string                    `json:"op"`
+	Gas           uint64                    `json:"gas"`
+	GasCost       uint64                    `json:"gasCost"`
+	Memory        []byte                    `json:"memory,omitempty"`
+	MemorySize    int                       `json:"memSize"`
+	Stack         []*big.Int                `json:"stack"`
+	ReturnData    []byte                    `json:"returnData,omitempty"`
+	Storage       map[types.Hash]types.Hash `json:"storage"`
+	Depth         int                       `json:"depth"`
+	RefundCounter uint64                    `json:"refund"`
+	Err           error                     `json:"err"`
+}
+
+func (l *StructLog) ErrorString() string {
+	if l.Err != nil {
+		return l.Err.Error()
+	}
+
+	return ""
+}
+
+type StructTracer struct {
+	Config Config
+
+	reason    error
+	interrupt uint32
+
+	logs        []StructLog
+	gasLimit    uint64
+	consumedGas uint64
+	output      []byte
+	err         error
+	storage     map[types.Address]map[types.Hash]types.Hash
+
+	currentMemory []byte
+	currentStack  []*big.Int
+}
+
+func NewStructTracer(config Config) *StructTracer {
+	return &StructTracer{
+		Config:  config,
+		storage: make(map[types.Address]map[types.Hash]types.Hash),
+	}
+}
+
+func (t *StructTracer) Cancel(err error) {
+	t.reason = err
```

Comment:
Why we don't need it trace safe?

Reply:
What do you mean? Did you mean you want me to use `atomic.Value` for this field?

Reply:
I don't think we need thread safety here since it's Cancel is only called from the timeoutCtx when creating the tracer, but let's make it thread safe as we might throw more `reason` errors in the future

Reply:
Fixed 9601ce9adfbd3a7d1022a643f797f0cc82f925b3